HIP/hip-336.md

```diff
@@ -0,0 +1,301 @@
+---
+hip: 336
+title: Approval and Allowance API for Tokens
+author: Albert Tam <albert.tam@hedera.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Review
+created: 2022-01-11
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/337
+updated: 2022-01-24
+---
+
+## Abstract
+
+Describes the addition of Services APIs to approve and exercise allowances to a third party account. The allowances grant another account to transfer hbar, fungible and non-fungible tokens from your account.
+
+## Motivation
+
+The current Services CryptoTransfer API allows a user to transfer hbar, fungible and non-fungible tokens but does not permit any other account to act as a delegate or proxy. The addition of the allowance APIs will permit cryptocurrency exchanges and wallets to trade on their customer's behalf without requiring the customer to  sign every transaction in advance.
+
+## User stories
+
+As a user, I want to enable another user the ability to spend `x` tokens (hbar, fungible and non-fungible tokens) from my account so that they can trade on my behalf without requiring that I expose my private key to them.
+
+As a user, I want to enable another user the ability to spend all instances of a particular NFT that I currently or may in the future hold in my account so that they can trade on my behalf without requiring that I expose my private key to them.
+
+As a user, I want to modify or remove an allowance I have granted to another user so that I can manage that user's ability to spend my tokens over time.
+
+As a user that has been granted an allowance, I want to be able to  transfer tokens owned by the owner account as if they were my tokens so that I can issue transfers on their behalf without them having to expose their private key.
+
+As a user, I want to query my account for the current token allowances that I have granted to other users so that I can monitor the activity of my delegates.
+  
+## Specification
+
+### Terminology
+
+The following terms will be used repeatedly throughout this section:
+
+- owner: the account that owns the tokens.
+- spender: the delegate account, the account being granted the power to spend the owner’s tokens.
+- recipient: the receiver of tokens when the spender issues a `transferFrom` operation.
+
+### Add Allowance Balances to CryptoGetInfoResponse
+
+The repeated fields `allowances` and `tokenAllowances` will be added to `CryptoGetInfoResponse` message. The `allowances` and `tokenAllowances` fields will contain hbar and token (fungible and non-fungible) allowances respectively.
+
+```protobuf
+// existing protobuf
+message CryptoGetInfoResponse {
+	// new nested message
+	message CryptoAllowance {
+		AccountID spender = 1;
+		int64 allowance = 2;
+	}
+
+	// new nested message
+	message TokenAllowance {
+		TokenID tokenId = 1;
+		AccountID spender = 2;
+		google.protobuf.Int64Value allowance = 3;
```

Comment:
CryptoApproveAllowanceTransactionBody refers to this as `amount` instead of `allowance`. Should we do the same here?

Reply:
Fixed